File: packages/snow_draw_arrow_core/lib/src/arrow_render_core.dart
import 'dart:math' as math;

import 'arrow_geom.dart';
import 'arrow_types.dart';

typedef ArrowEndpointPosition = String;

const arrowEndpointPositionStart = 'start';
const arrowEndpointPositionEnd = 'end';

class ArrowheadPointsInput {
  const ArrowheadPointsInput({
    required this.arrowPoints,
    required this.strokeWidth,
    required this.curveOps,
    required this.position,
    required this.arrowhead,
  });

  final List<Point> arrowPoints;
  final double strokeWidth;
  final List<CurvePathOp> curveOps;
  final ArrowEndpointPosition position;
  final Arrowhead arrowhead;
}

class ArrowheadRenderPrimitivesInput extends ArrowheadPointsInput {
  const ArrowheadRenderPrimitivesInput({
    required super.arrowPoints,
    required super.strokeWidth,
    required super.curveOps,
    required super.position,
    required super.arrowhead,
    required this.strokeStyle,
  });

  final ArrowStrokeStyle strokeStyle;
}

sealed class ArrowheadRenderPrimitive {
  const ArrowheadRenderPrimitive({
    required this.kind,
    required this.dashMode,
    this.roughnessCap,
  });

  final String kind;
  final ArrowheadDashMode dashMode;
  final double? roughnessCap;
}

class ArrowheadLinePrimitive extends ArrowheadRenderPrimitive {
  const ArrowheadLinePrimitive({
    required this.from,
    required this.to,
    required super.dashMode,
    super.roughnessCap,
  }) : super(kind: 'line');

  final Point from;
  final Point to;
}

class ArrowheadPolygonPrimitive extends ArrowheadRenderPrimitive {
  const ArrowheadPolygonPrimitive({
    required this.points,
    required this.fillMode,
    required super.roughnessCap,
  }) : super(kind: 'polygon', dashMode: 'solid');

  final List<Point> points;
  final ArrowheadFillMode fillMode;
}

class ArrowheadCirclePrimitive extends ArrowheadRenderPrimitive {
  const ArrowheadCirclePrimitive({
    required this.center,
    required this.diameter,
    required this.fillMode,
    required super.roughnessCap,
  }) : super(kind: 'circle', dashMode: 'solid');

  final Point center;
  final double diameter;
  final ArrowheadFillMode fillMode;
}

double getArrowheadSize(Arrowhead arrowhead) {
  switch (arrowhead) {
    case 'arrow':
      return 25;
    case 'diamond':
    case 'diamond_outline':
      return 12;
    case 'crowfoot_many':
    case 'crowfoot_one':
    case 'crowfoot_one_or_many':
      return 20;
    default:
      return 15;
  }
}

double getArrowheadAngle(Arrowhead arrowhead) {
  switch (arrowhead) {
    case 'bar':
      return 90;
    case 'arrow':
      return 20;
    default:
      return 25;
  }
}

double _toRadians(double degrees) => (degrees * math.pi) / 180;

bool _isCurveOp(CurvePathOp? op) => op != null;

Point _getCurvePoint(CurvePathOp op, int xIndex, int yIndex) => <double>[
  op.data[xIndex],
  op.data[yIndex],
];

Point? _getCurveStartPoint(CurvePathOp? previousOp) {
  if (previousOp == null) {
    return null;
  }
  if (previousOp.op == 'move' && previousOp.data.length >= 2) {
    return <double>[previousOp.data[0], previousOp.data[1]];
  }
  if (previousOp.op == 'bcurveTo' && previousOp.data.length >= 6) {
    return <double>[previousOp.data[4], previousOp.data[5]];
  }
  return <double>[0, 0];
}

Point? _normalizeDirection(Point from, Point to) {
  final dx = to[0] - from[0];
  final dy = to[1] - from[1];
  final length = math.sqrt(dx * dx + dy * dy);
  if (length <= 1e-6) {
    return null;
  }
  return <double>[dx / length, dy / length];
}

Point _pointAtBezier(double t, Point p0, Point p1, Point p2, Point p3) {
  final oneMinusT = 1 - t;
  final oneMinusTSquared = oneMinusT * oneMinusT;
  final tSquared = t * t;
  final x =
      oneMinusT * oneMinusTSquared * p3[0] +
      3 * t * oneMinusTSquared * p2[0] +
      3 * tSquared * oneMinusT * p1[0] +
      p0[0] * tSquared * t;
  final y =
      oneMinusT * oneMinusTSquared * p3[1] +
      3 * t * oneMinusTSquared * p2[1] +
      3 * tSquared * oneMinusT * p1[1] +
      p0[1] * tSquared * t;
  return <double>[x, y];
}

double _getSegmentLength(
  List<Point> arrowPoints,
  ArrowEndpointPosition position,
) {
  if (arrowPoints.isEmpty) {
    return 0;
  }
  final currentPoint = position == arrowEndpointPositionEnd
      ? arrowPoints[arrowPoints.length - 1]
      : arrowPoints[0];
  final previousPoint = arrowPoints.length > 1
      ? position == arrowEndpointPositionEnd
            ? arrowPoints[arrowPoints.length - 2]
            : arrowPoints[1]
      : <double>[0, 0];
  return distance(currentPoint, previousPoint);
}

ArrowheadPoints? getArrowheadPoints(ArrowheadPointsInput input) {
  final arrowPoints = input.arrowPoints;
  final strokeWidth = input.strokeWidth;
  final curveOps = input.curveOps;
  final position = input.position;
  final arrowhead = input.arrowhead;

  if (arrowPoints.isEmpty || curveOps.length < 2) {
    return null;
  }

  final index = position == arrowEndpointPositionStart
      ? 1
      : curveOps.length - 1;
  final curveOp = index >= 0 && index < curveOps.length
      ? curveOps[index]
      : null;
  if (!_isCurveOp(curveOp) || curveOp!.data.length != 6) {
    return null;
  }

  final previousOp = curveOps[index - 1];
  final p0 = _getCurveStartPoint(previousOp);
  if (p0 == null) {
    return null;
  }

  final p1 = _getCurvePoint(curveOp, 0, 1);
  final p2 = _getCurvePoint(curveOp, 2, 3);
  final p3 = _getCurvePoint(curveOp, 4, 5);

  final endpoint = position == arrowEndpointPositionStart ? p0 : p3;
  final samplePoint = _pointAtBezier(0.3, p0, p1, p2, p3);
  final direction = _normalizeDirection(samplePoint, endpoint);
  if (direction == null) {
    return null;
  }

  final size = getArrowheadSize(arrowhead);
  final length = _getSegmentLength(arrowPoints, position);
  final lengthMultiplier =
      arrowhead == 'diamond' || arrowhead == 'diamond_outline' ? 0.25 : 0.5;
  final minSize = math.min(size, length * lengthMultiplier);
  final x2 = endpoint[0];
  final y2 = endpoint[1];
  final xs = x2 - direction[0] * minSize;
  final ys = y2 - direction[1] * minSize;

  if (arrowhead == 'dot' ||
      arrowhead == 'circle' ||
      arrowhead == 'circle_outline') {
    final diameter =
        math.sqrt((ys - y2) * (ys - y2) + (xs - x2) * (xs - x2)) +
        strokeWidth -
        2;
    return <double>[x2, y2, diameter];
  }

  final angle = getArrowheadAngle(arrowhead);
  if (arrowhead == 'crowfoot_many' || arrowhead == 'crowfoot_one_or_many') {
    final rotatedLeft = rotatePoint(
      <double>[x2, y2],
      <double>[xs, ys],
      _toRadians(-angle),
    );
    final rotatedRight = rotatePoint(
      <double>[x2, y2],
      <double>[xs, ys],
      _toRadians(angle),
    );
    return <double>[
      xs,
      ys,
      rotatedLeft[0],
      rotatedLeft[1],
      rotatedRight[0],
      rotatedRight[1],
    ];
  }

  final rotatedLeft = rotatePoint(
    <double>[xs, ys],
    <double>[x2, y2],
    _toRadians(-angle),
  );
  final rotatedRight = rotatePoint(
    <double>[xs, ys],
    <double>[x2, y2],
    _toRadians(angle),
  );
  final x3 = rotatedLeft[0];
  final y3 = rotatedLeft[1];
  final x4 = rotatedRight[0];
  final y4 = rotatedRight[1];

  if (arrowhead == 'diamond' || arrowhead == 'diamond_outline') {
    final previousPoint = position == arrowEndpointPositionStart
        ? arrowPoints.length > 1
              ? arrowPoints[1]
              : <double>[0, 0]
        : arrowPoints.length > 1
        ? arrowPoints[arrowPoints.length - 2]
        : <double>[0, 0];
    final oppositeSeed = position == arrowEndpointPositionStart
        ? <double>[x2 + minSize * 2, y2]
        : <double>[x2 - minSize * 2, y2];
    final oppositeAngle = position == arrowEndpointPositionStart
        ? math.atan2(previousPoint[1] - y2, previousPoint[0] - x2)
        : math.atan2(y2 - previousPoint[1], x2 - previousPoint[0]);
    final oppositePoint = rotatePoint(oppositeSeed, <double>[
      x2,
      y2,
    ], oppositeAngle);

    return <double>[x2, y2, x3, y3, oppositePoint[0], oppositePoint[1], x4, y4];
  }

  return <double>[x2, y2, x3, y3, x4, y4];
}

Point _toPoint(double x, double y) => <double>[x, y];

List<ArrowheadRenderPrimitive> getArrowheadRenderPrimitives(
  ArrowheadRenderPrimitivesInput input,
) {
  final points = getArrowheadPoints(input);
  if (points == null) {
    return <ArrowheadRenderPrimitive>[];
  }

  ArrowheadLinePrimitive line(
    Point from,
    Point to,
    ArrowheadDashMode dashMode, [
    double? roughnessCap,
  ]) => ArrowheadLinePrimitive(
    from: from,
    to: to,
    dashMode: dashMode,
    roughnessCap: roughnessCap,
  );

  ArrowheadPolygonPrimitive polygon(
    List<Point> polygonPoints,
    ArrowheadFillMode fillMode,
    double roughnessCap,
  ) => ArrowheadPolygonPrimitive(
    points: polygonPoints,
    fillMode: fillMode,
    roughnessCap: roughnessCap,
  );

  ArrowheadCirclePrimitive circle(
    Point center,
    double diameter,
    ArrowheadFillMode fillMode,
    double roughnessCap,
  ) => ArrowheadCirclePrimitive(
    center: center,
    diameter: diameter,
    fillMode: fillMode,
    roughnessCap: roughnessCap,
  );

  switch (input.arrowhead) {
    case 'dot':
    case 'circle':
      return <ArrowheadRenderPrimitive>[
        circle(_toPoint(points[0], points[1]), points[2], 'stroke', 0.5),
      ];
    case 'circle_outline':
      return <ArrowheadRenderPrimitive>[
        circle(_toPoint(points[0], points[1]), points[2], 'background', 0.5),
      ];
    case 'triangle':
    case 'triangle_outline':
      if (points.length != 6) {
        return <ArrowheadRenderPrimitive>[];
      }
      final x = points[0];
      final y = points[1];
      final x2 = points[2];
      final y2 = points[3];
      final x3 = points[4];
      final y3 = points[5];
      return <ArrowheadRenderPrimitive>[
        polygon(
          <Point>[
            _toPoint(x, y),
            _toPoint(x2, y2),
            _toPoint(x3, y3),
            _toPoint(x, y),
          ],
          input.arrowhead == 'triangle_outline' ? 'background' : 'stroke',
          1,
        ),
      ];
    case 'diamond':
    case 'diamond_outline':
      if (points.length != 8) {
        return <ArrowheadRenderPrimitive>[];
      }
      final x = points[0];
      final y = points[1];
      final x2 = points[2];
      final y2 = points[3];
      final x3 = points[4];
      final y3 = points[5];
      final x4 = points[6];
      final y4 = points[7];
      return <ArrowheadRenderPrimitive>[
        polygon(
          <Point>[
            _toPoint(x, y),
            _toPoint(x2, y2),
            _toPoint(x3, y3),
            _toPoint(x4, y4),
            _toPoint(x, y),
          ],
          input.arrowhead == 'diamond_outline' ? 'background' : 'stroke',
          1,
        ),
      ];
    case 'crowfoot_one':
      if (points.length != 6) {
        return <ArrowheadRenderPrimitive>[];
      }
      return <ArrowheadRenderPrimitive>[
        line(
          _toPoint(points[2], points[3]),
          _toPoint(points[4], points[5]),
          'inherit',
        ),
      ];
    case 'bar':
    case 'arrow':
    case 'crowfoot_many':
    case 'crowfoot_one_or_many':
    default:
      if (points.length != 6) {
        return <ArrowheadRenderPrimitive>[];
      }
      final x2 = points[0];
      final y2 = points[1];
      final x3 = points[2];
      final y3 = points[3];
      final x4 = points[4];
      final y4 = points[5];
      final dashMode = input.strokeStyle == 'dotted' ? 'dotted-cap' : 'solid';
      final primitives = <ArrowheadRenderPrimitive>[
        line(_toPoint(x3, y3), _toPoint(x2, y2), dashMode, 1),
        line(_toPoint(x4, y4), _toPoint(x2, y2), dashMode, 1),
      ];

      if (input.arrowhead == 'crowfoot_one_or_many') {
        final crowfootOnePoints = getArrowheadPoints(
          ArrowheadPointsInput(
            arrowPoints: input.arrowPoints,
            strokeWidth: input.strokeWidth,
            curveOps: input.curveOps,
            position: input.position,
            arrowhead: 'crowfoot_one',
          ),
        );
        if (crowfootOnePoints != null && crowfootOnePoints.length == 6) {
          primitives.add(
            line(
              _toPoint(crowfootOnePoints[2], crowfootOnePoints[3]),
              _toPoint(crowfootOnePoints[4], crowfootOnePoints[5]),
              dashMode,
              1,
            ),
          );
        }
      }

      return primitives;
  }
}

bool _isHorizontal(Point a, Point b) =>
    (a[0] - b[0]).abs() > (a[1] - b[1]).abs();

String _formatPathNumber(double value) {
  if (value == value.truncateToDouble()) {
    return value.toInt().toString();
  }
  return value.toString();
}

String generateElbowArrowPath(List<Point> points, double radius) {
  if (points.isEmpty) {
    return '';
  }
  if (points.length == 1) {
    return 'M ${_formatPathNumber(points[0][0])} '
        '${_formatPathNumber(points[0][1])}';
  }

  final subpoints = <Point>[];
  for (var i = 1; i < points.length - 1; i += 1) {
    final previous = points[i - 1];
    final next = points[i + 1];
    final point = points[i];
    final prevIsHorizontal = _isHorizontal(point, previous);
    final nextIsHorizontal = _isHorizontal(next, point);
    final corner = math.min(
      radius,
      math.min(
        distance(points[i], next) / 2,
        distance(points[i], previous) / 2,
      ),
    );

    if (prevIsHorizontal) {
      subpoints.add(<double>[
        if (previous[0] < point[0]) point[0] - corner else point[0] + corner,
        point[1],
      ]);
    } else {
      subpoints.add(<double>[
        point[0],
        if (previous[1] < point[1]) point[1] - corner else point[1] + corner,
      ]);
    }

    subpoints.add(point);

    if (nextIsHorizontal) {
      subpoints.add(<double>[
        if (next[0] < point[0]) point[0] - corner else point[0] + corner,
        point[1],
      ]);
    } else {
      subpoints.add(<double>[
        point[0],
        if (next[1] < point[1]) point[1] - corner else point[1] + corner,
      ]);
    }
  }

  final path = <String>[
    'M ${_formatPathNumber(points[0][0])} ${_formatPathNumber(points[0][1])}',
  ];
  for (var i = 0; i < subpoints.length; i += 3) {
    path
      ..add(
        'L ${_formatPathNumber(subpoints[i][0])} '
        '${_formatPathNumber(subpoints[i][1])}',
      )
      ..add(
        'Q ${_formatPathNumber(subpoints[i + 1][0])} '
        '${_formatPathNumber(subpoints[i + 1][1])}, '
        '${_formatPathNumber(subpoints[i + 2][0])} '
        '${_formatPathNumber(subpoints[i + 2][1])}',
      );
  }
  path.add(
    'L ${_formatPathNumber(points[points.length - 1][0])} '
    '${_formatPathNumber(points[points.length - 1][1])}',
  );
  return path.join(' ');
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/core/arrow_render_core.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;

use super::arrow_types::{
    ArrowStrokeStyle, Arrowhead, ArrowheadDashMode, ArrowheadFillMode, CurvePathOp,
};

pub type ArrowEndpointPosition = &'static str;

pub const ARROW_ENDPOINT_POSITION_START: ArrowEndpointPosition = "start";
pub const ARROW_ENDPOINT_POSITION_END: ArrowEndpointPosition = "end";

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowheadPointsInput {
    pub arrow_points: Vec<DrawPoint>,
    pub stroke_width: f64,
    pub curve_ops: Vec<CurvePathOp>,
    pub position: ArrowEndpointPosition,
    pub arrowhead: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowheadRenderPrimitivesInput {
    pub arrow_points: Vec<DrawPoint>,
    pub stroke_width: f64,
    pub curve_ops: Vec<CurvePathOp>,
    pub position: ArrowEndpointPosition,
    pub arrowhead: String,
    pub stroke_style: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowheadLinePrimitive {
    pub from: DrawPoint,
    pub to: DrawPoint,
    pub dash_mode: String,
    pub roughness_cap: Option<f64>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowheadPolygonPrimitive {
    pub points: Vec<DrawPoint>,
    pub fill_mode: String,
    pub roughness_cap: Option<f64>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArrowheadCirclePrimitive {
    pub center: DrawPoint,
    pub diameter: f64,
    pub fill_mode: String,
    pub roughness_cap: Option<f64>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum ArrowheadRenderPrimitive {
    Line {
        from: DrawPoint,
        to: DrawPoint,
        dash_mode: String,
        roughness_cap: Option<f64>,
    },
    Polygon {
        points: Vec<DrawPoint>,
        fill_mode: String,
        roughness_cap: Option<f64>,
    },
    Circle {
        center: DrawPoint,
        diameter: f64,
        fill_mode: String,
        roughness_cap: Option<f64>,
    },
}

impl From<ArrowheadLinePrimitive> for ArrowheadRenderPrimitive {
    fn from(value: ArrowheadLinePrimitive) -> Self {
        Self::Line {
            from: value.from,
            to: value.to,
            dash_mode: value.dash_mode,
            roughness_cap: value.roughness_cap,
        }
    }
}

impl From<ArrowheadPolygonPrimitive> for ArrowheadRenderPrimitive {
    fn from(value: ArrowheadPolygonPrimitive) -> Self {
        Self::Polygon {
            points: value.points,
            fill_mode: value.fill_mode,
            roughness_cap: value.roughness_cap,
        }
    }
}

impl From<ArrowheadCirclePrimitive> for ArrowheadRenderPrimitive {
    fn from(value: ArrowheadCirclePrimitive) -> Self {
        Self::Circle {
            center: value.center,
            diameter: value.diameter,
            fill_mode: value.fill_mode,
            roughness_cap: value.roughness_cap,
        }
    }
}

pub fn get_arrowhead_size(arrowhead: &str) -> f64 {
    match arrowhead {
        "arrow" => 25.0,
        "diamond" | "diamond_outline" => 12.0,
        "crowfoot_many" | "crowfoot_one" | "crowfoot_one_or_many" => 20.0,
        _ => 15.0,
    }
}

pub fn get_arrowhead_angle(arrowhead: &str) -> f64 {
    match arrowhead {
        "bar" => 90.0,
        "arrow" => 20.0,
        _ => 25.0,
    }
}

pub fn get_arrowhead_points(input: &ArrowheadPointsInput) -> Option<Vec<f64>> {
    let arrow_points = &input.arrow_points;
    if arrow_points.len() < 2 || input.stroke_width <= 0.0 {
        return None;
    }

    let is_end = input.position == ARROW_ENDPOINT_POSITION_END;
    let tip = if is_end {
        *arrow_points.last()?
    } else {
        arrow_points[0]
    };
    let neighbor = if is_end {
        arrow_points[arrow_points.len() - 2]
    } else {
        arrow_points[1]
    };

    let direction = normalize_direction(neighbor, tip)?;
    let size = get_arrowhead_size(&input.arrowhead) * (input.stroke_width / 2.0).max(1.0) / 2.0;
    let half = size / 2.0;
    let angle = get_arrowhead_angle(&input.arrowhead).to_radians();
    let shaft = DrawPoint::new(-direction.x, -direction.y);
    let perpendicular = DrawPoint::new(-shaft.y, shaft.x);

    let base_center = DrawPoint::new(tip.x + shaft.x * size, tip.y + shaft.y * size);
    let left = rotate_around(base_center + perpendicular * half, tip, angle);
    let right = rotate_around(base_center - perpendicular * half, tip, -angle);

    match input.arrowhead.as_str() {
        "dot" | "circle" | "circle_outline" => Some(vec![tip.x, tip.y, size]),
        "square" => {
            let c1 = base_center + perpendicular * half + shaft * half;
            let c2 = base_center - perpendicular * half + shaft * half;
            let c3 = base_center - perpendicular * half - shaft * half;
            let c4 = base_center + perpendicular * half - shaft * half;
            Some(vec![c1.x, c1.y, c2.x, c2.y, c3.x, c3.y, c4.x, c4.y])
        }
        "diamond" | "diamond_outline" => {
            let base = DrawPoint::new(tip.x + shaft.x * size, tip.y + shaft.y * size);
            let back = DrawPoint::new(tip.x + shaft.x * size * 2.0, tip.y + shaft.y * size * 2.0);
            let left_mid = base + perpendicular * half;
            let right_mid = base - perpendicular * half;
            Some(vec![
                tip.x,
                tip.y,
                left_mid.x,
                left_mid.y,
                back.x,
                back.y,
                right_mid.x,
                right_mid.y,
            ])
        }
        _ => Some(vec![tip.x, tip.y, left.x, left.y, right.x, right.y]),
    }
}

pub fn get_arrowhead_render_primitives(
    input: &ArrowheadRenderPrimitivesInput,
) -> Vec<ArrowheadRenderPrimitive> {
    let Some(points) = get_arrowhead_points(&ArrowheadPointsInput {
        arrow_points: input.arrow_points.clone(),
        stroke_width: input.stroke_width,
        curve_ops: input.curve_ops.clone(),
        position: input.position,
        arrowhead: input.arrowhead.clone(),
    }) else {
        return Vec::new();
    };

    match input.arrowhead.as_str() {
        "dot" => vec![ArrowheadRenderPrimitive::Circle {
            center: DrawPoint::new(points[0], points[1]),
            diameter: points[2],
            fill_mode: "stroke".to_string(),
            roughness_cap: Some(1.0),
        }],
        "circle" | "circle_outline" => vec![ArrowheadRenderPrimitive::Circle {
            center: DrawPoint::new(points[0], points[1]),
            diameter: points[2],
            fill_mode: if input.arrowhead == "circle_outline" {
                "background".to_string()
            } else {
                "stroke".to_string()
            },
            roughness_cap: Some(1.0),
        }],
        "square" => vec![ArrowheadRenderPrimitive::Polygon {
            points: vec![
                DrawPoint::new(points[0], points[1]),
                DrawPoint::new(points[2], points[3]),
                DrawPoint::new(points[4], points[5]),
                DrawPoint::new(points[6], points[7]),
            ],
            fill_mode: "stroke".to_string(),
            roughness_cap: Some(1.0),
        }],
        "diamond" | "diamond_outline" => vec![ArrowheadRenderPrimitive::Polygon {
            points: vec![
                DrawPoint::new(points[0], points[1]),
                DrawPoint::new(points[2], points[3]),
                DrawPoint::new(points[4], points[5]),
                DrawPoint::new(points[6], points[7]),
                DrawPoint::new(points[0], points[1]),
            ],
            fill_mode: if input.arrowhead == "diamond_outline" {
                "background".to_string()
            } else {
                "stroke".to_string()
            },
            roughness_cap: Some(1.0),
        }],
        _ => {
            if points.len() != 6 {
                return Vec::new();
            }
            let dash_mode = if input.stroke_style == "dotted" {
                "dotted-cap"
            } else {
                "solid"
            };
            vec![
                ArrowheadRenderPrimitive::Line {
                    from: DrawPoint::new(points[2], points[3]),
                    to: DrawPoint::new(points[0], points[1]),
                    dash_mode: dash_mode.to_string(),
                    roughness_cap: Some(1.0),
                },
                ArrowheadRenderPrimitive::Line {
                    from: DrawPoint::new(points[4], points[5]),
                    to: DrawPoint::new(points[0], points[1]),
                    dash_mode: dash_mode.to_string(),
                    roughness_cap: Some(1.0),
                },
            ]
        }
    }
}

pub fn generate_elbow_arrow_path(points: &[DrawPoint], radius: f64) -> String {
    if points.is_empty() {
        return String::new();
    }
    if points.len() == 1 {
        return format!(
            "M {} {}",
            format_number(points[0].x),
            format_number(points[0].y)
        );
    }

    let mut subpoints = Vec::<DrawPoint>::new();
    for index in 1..points.len() - 1 {
        let previous = points[index - 1];
        let point = points[index];
        let next = points[index + 1];
        let prev_is_horizontal = (point.x - previous.x).abs() > (point.y - previous.y).abs();
        let next_is_horizontal = (next.x - point.x).abs() > (next.y - point.y).abs();
        let corner = radius
            .min(point.distance(next) / 2.0)
            .min(point.distance(previous) / 2.0);

        subpoints.push(if prev_is_horizontal {
            DrawPoint::new(
                if previous.x < point.x {
                    point.x - corner
                } else {
                    point.x + corner
                },
                point.y,
            )
        } else {
            DrawPoint::new(
                point.x,
                if previous.y < point.y {
                    point.y - corner
                } else {
                    point.y + corner
                },
            )
        });
        subpoints.push(point);
        subpoints.push(if next_is_horizontal {
            DrawPoint::new(
                if next.x < point.x {
                    point.x - corner
                } else {
                    point.x + corner
                },
                point.y,
            )
        } else {
            DrawPoint::new(
                point.x,
                if next.y < point.y {
                    point.y - corner
                } else {
                    point.y + corner
                },
            )
        });
    }

    let mut path = vec![format!(
        "M {} {}",
        format_number(points[0].x),
        format_number(points[0].y)
    )];
    let mut index = 0usize;
    while index + 2 < subpoints.len() {
        path.push(format!(
            "L {} {}",
            format_number(subpoints[index].x),
            format_number(subpoints[index].y)
        ));
        path.push(format!(
            "Q {} {}, {} {}",
            format_number(subpoints[index + 1].x),
            format_number(subpoints[index + 1].y),
            format_number(subpoints[index + 2].x),
            format_number(subpoints[index + 2].y)
        ));
        index += 3;
    }
    if let Some(last) = points.last() {
        path.push(format!(
            "L {} {}",
            format_number(last.x),
            format_number(last.y)
        ));
    }
    path.join(" ")
}

fn normalize_direction(from: DrawPoint, to: DrawPoint) -> Option<DrawPoint> {
    let dx = to.x - from.x;
    let dy = to.y - from.y;
    let length = (dx * dx + dy * dy).sqrt();
    (length > 1e-6).then_some(DrawPoint::new(dx / length, dy / length))
}

fn rotate_around(point: DrawPoint, center: DrawPoint, angle: f64) -> DrawPoint {
    let translated_x = point.x - center.x;
    let translated_y = point.y - center.y;
    let cos = angle.cos();
    let sin = angle.sin();
    DrawPoint::new(
        center.x + translated_x * cos - translated_y * sin,
        center.y + translated_x * sin + translated_y * cos,
    )
}

fn format_number(value: f64) -> String {
    if value == value.trunc() {
        (value as i64).to_string()
    } else {
        value.to_string()
    }
}
